### PR TITLE
Add upsert functionality to allow company updates to be added to alpha search

### PR DESCRIPTION
**Pull request by caismith-ch**
Add functionality to upsert a company to the alpha search index. Upsert is an elasticsearch term to update or create a document depending on whether it exists or not.

- POST request to create/update a company in alpha_search index
- Add wrapped UpsertException
- Add JUnit tests for all new code
- Increase overall test coverage
- Update SearchRestClient to be more generic by calling it RestClientService and adding upsert method

**Resolves**: PCI-415

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/search/api/service/upsert/UpsertRequestService.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    public IndexRequest createIndexRequest(Company company) throws UpsertException {
+
+        IndexRequest indexRequest;
+        try {
+            LOG.info("Preparing index request for "  + company.getId());
+            indexRequest = new IndexRequest("alpha_search")
+                .source(buildRequest(company));
+            return indexRequest;
+        } catch (IOException e) {
+            LOG.error("Failed to index a document");
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/search/api/service/upsert/UpsertRequestService.java]: Can we have something to identify which document failed to index?

File: src/main/java/uk/gov/companieshouse/search/api/service/upsert/UpsertRequestService.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        UpdateRequest updateRequest;
+        try {
+            LOG.info("Attempt to upsert document if it does not exist for "  + company.getId());
+            updateRequest = new UpdateRequest("alpha_search", company.getId())
+                .doc(buildRequest(company))
+                .upsert(indexRequest);
+
+            return updateRequest;
+        } catch (IOException e) {
+            LOG.error("Failed to upsert document");
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/search/api/service/upsert/UpsertRequestService.java]: same as above regarding identification of document that failed

File: src/main/java/uk/gov/companieshouse/search/api/mapper/ApiToResponseMapper.java [hunk 3]
```diff
@@ -18,6 +19,8 @@ public ResponseEntity map(ResponseObject responseObject) {
                 return ResponseEntity.status(FOUND).body(responseObject.getData());
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/search/api/mapper/ApiToResponseMapper.java]: where is the apiToReponseMapper elements for UPDATE_REQUEST_ERROR and UPSERT_ERROR?

Comment (caismith-ch) [reply src/main/java/uk/gov/companieshouse/search/api/mapper/ApiToResponseMapper.java]: Good spot must have missed them.

File: src/main/java/uk/gov/companieshouse/search/api/mapper/ApiToResponseMapper.java [hunk 4]
```diff
@@ -18,6 +20,12 @@ public ResponseEntity map(ResponseObject responseObject) {
...
             case SEARCH_NOT_FOUND:
                 return ResponseEntity.status(NOT_FOUND).build();
+            case DOCUMENT_UPSERTED:
+                return ResponseEntity.status(OK).build();
+            case UPDATE_REQUEST_ERROR:
+                return ResponseEntity.status(BAD_REQUEST).build();
+            case UPSERT_ERROR:
+                return ResponseEntity.status(BAD_REQUEST).build();
```
Comment (Mathsh) [inline src/main/java/uk/gov/companieshouse/search/api/mapper/ApiToResponseMapper.java]: this 
`case UPDATE_REQUEST_ERROR:
                return ResponseEntity.status(BAD_REQUEST).build();
            case UPSERT_ERROR:
                return ResponseEntity.status(BAD_REQUEST).build();`

Can be written as 
          `case UPDATE_REQUEST_ERROR:
            case UPSERT_ERROR:
                return ResponseEntity.status(BAD_REQUEST).build();`

**Pull request merged**